specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,988 @@
+
+; module flags
+
+!llvm.module.flags = !{!0, !1, !2, !3}
+
+!0 = !{i32 1, !"qir_major_version", i32 1}
+!1 = !{i32 7, !"qir_minor_version", i32 0}
+!2 = !{i32 1, !"dynamic_qubit_management", i1 false}
+!3 = !{i32 1, !"dynamic_result_management", i1 false}
+!4 = !{i32 1, !"qubit_resetting", i1 true} ; bullet 8 qubit resetting
+!5 = !{i32 1, !"classical_ints", i1 false} ; ...
+!6 = !{i32 1, !"classical_floats", i1 false}
+!7 = !{i32 1, !"classical_fixed_points", i1 false}
+!8 = !{i32 1, !"user_functions", i1 false}
+!9 = !{i32 1, !"dynamic_float_args", i1 false}
+!10 = !{i32 1, !"extern_functions", i1 false}
+!11 = !{i32 1, !"backwards_branching", i1 false}
+```
+
+The program performs gate teleportation, and it uses conditional single qubit gates and mid-circuit measurements to effect control flow.
+
+For the sake of clarity, the code above does not contain any [debug
+symbols](https://releases.llvm.org/13.0.0/docs/tutorial/MyFirstLanguageFrontend/LangImpl09.html?highlight=debug%20symbols).
+Debug symbols contain information that is used by a debugger to relate failures
+during execution back to the original source code. While we expect this to be
+primarily useful for execution on simulators, debug symbols are both easy to
+ignore and may be useful to generate helpful error messages for compilation
+failures that happen only late in the process. We hence see no reason to
+disallow them from occurring in the bitcode but will not detail their use any
+further. We defer to existing resources for more information about how to
+generate and use debug symbols.
+
+## Entry Point Definition
+
+The bitcode contains the definition of the LLVM function that should be invoked
+when the program is executed, referred to as "entry point" in the rest of this
+profile specification. The name of this function may be chosen freely, as long
+as it is a valid [global
+identifier](https://llvm.org/docs/LangRef.html#identifiers) according to the
+LLVM standard. The entry point is identified by a custom function attribute; the
+section on [attributes](#attributes) defines which attributes must be attached
+to the entry point function.
+
+An entry point function may not take any parameters and must either return void in case the program has no failure conditions or must return an exit code in the
+form of a 64-bit integer. The exit code `0` must be used to indicate a
+successful execution of the program. Exit code `1` can be used to illustrate
+that a real-time error occurred in the execution of the program dependent
+upon what features a back-end opts into when supporting an Adaptive Profile program.
+For example, with **Bullet 9/14** classical code may encounter errors like division
+by zero that can cause an entry-point function to fail. In this case, programs must dynamically check failure conditions and return the exit code instead of failing outright.
```

Comment:
I think this point bears more discussion. The entry-point function is intended to have different return values for indicating shot success, and it would be nice if a shot always has an associated return value to indicate whether a shot finished successfully or failed due to the user program expressing a real-time classical logic failure. I do think that as the profiles become more expressive this should be less constrained.

Reply:
I do think I can be talked off the Adaptive Profile programs must have checked divisions point, however, if that's a broadly shared sentiment. Admittedly, putting that constraint on an IR seems aggressive.

Reply:
It's less about the divide-by-zero, it's just a bit more foundational than that. Say you have some C# code that calls Q# which got compiled into QIR and sent to be run on a remote QPU. The 'result' of that entry point when it's run is your results dictionary built up via the various intrisics, _not_ the error code. Even if your execution failed, you wouldn't just get back an error code, you'd get an actual exception in your language of choice. In both cases the IR itself is not representative of what is actually being done. 

It also breaks down when you have the concept of far more complex code which actually does serious post-processing in the QIR itself and returns the results of that, which will be a clasical value. In that situation you simply can't have this, because the return from the entry-point method is, well, the actual return value. (Adding infinite `record_x` intrinsics is not a nice solution.)

Just switch the solutions. Leave the flow in the IR as you'd expect and just have an intrinsic that's `qis_rt_return_code(code)` or something if you want to be able to throw specific errors with codes.

Reply:
I think this boils down to whether we require a single return or not. We think at its core, this should not be required, and will make support for multiple returns an optional feature. We will still require the return code to be conveyed as an actual return statement; reasonably, anything that requires giving a non-zero return code also requires to have support for one or more classical data types, making it possible to propagate the code to the final return.